src/main.rs

```diff
@@ -30,49 +30,52 @@ mod overhead_light;
 mod rgb_led;
 mod serial;
 
+static mut USB_ENDPOINT_MEMORY: [u32; 1024] = [0; 1024];
+
 #[entry]
 fn main() -> ! {
-    let mut cp =
-        cortex_m::peripheral::Peripherals::take().expect("failed to get cortex_m peripherals");
-    let dp = pac::Peripherals::take().expect("failed to get stm32 peripherals");
+    let cp = cortex_m::peripheral::Peripherals::take().expect("failed to get cortex_m peripherals");
+    let dp = stm32::Peripherals::take().expect("failed to get stm32 peripherals");
 
-    // Take ownership over the raw flash and rcc devices and convert them into the corresponding
+    // Take ownership over the raw devices and convert them into the corresponding
     // HAL structs.
     // RCC = Reset and Clock Control
-    let mut flash = dp.FLASH.constrain();
-    let mut rcc = dp.RCC.constrain();
+    let rcc = dp.RCC.constrain();
 
     // The various system clocks need to be configured to particular values
     // to work with USB - we'll set them up here.
     let clocks = rcc
         .cfgr
-        .use_hse(8.mhz()) // Use the High Speed External 8MHz crystal
+        .use_hse(25.mhz()) // Use the High Speed External 25MHz crystal
         .sysclk(48.mhz()) // The main system clock will be 48MHz
-        .pclk1(24.mhz())  // Use 24MHz for the APB1 (Advanced Peripheral Bus 1)
-        .freeze(&mut flash.acr);
-
-    assert!(clocks.usbclk_valid());
-
-    // Needed in order for MonoTimer to work properly
-    cp.DCB.enable_trace();
+        .require_pll48clk()
+        .freeze();
 
-    // Prepare the alternate function I/O registers
-    let mut afio = dp.AFIO.constrain(&mut rcc.apb2);
+    // assert!(clocks.usbclk_valid());
 
     // Grab the GPIO banks we'll use.
-    let mut gpioa = dp.GPIOA.split(&mut rcc.apb2);
-    let mut gpiob = dp.GPIOB.split(&mut rcc.apb2);
+    let gpioa = dp.GPIOA.split();
+    let gpiob = dp.GPIOB.split();
+    let gpioc = dp.GPIOC.split();
 
-    // Set up the LED (B12).
-    let mut led = gpiob.pb12.into_push_pull_output(&mut gpiob.crh);
+    // Set up the LED (C13).
+    let mut led = gpioc.pc13.into_push_pull_output();
 
     // Set up USB communications
-    let usb_pin_d_plus = gpioa.pa12;
-    let usb_pin_d_minus = gpioa.pa11;
+    let usb_pin_d_plus = gpioa.pa12.into_alternate_af10();
```

Comment:
it was not really obvious what alternate function 10 does, reading a few docs. is this a way to map a GPIO pin to a different functionality? and if so, where is it documented what function each `afn` corresponds to?

Reply:
Check out pages 47-52 in the data sheet:

https://www.st.com/resource/en/datasheet/stm32f411re.pdf

It's also encoded into the type system and you can see the various combinations in the docs (click on the `PinC1`, `PinC2`, etc. traits, find the timer it's generic over, then see which pins implement that trait)

https://docs.rs/stm32f4xx-hal/0.8.3/stm32f4xx_hal/timer/trait.PinC1.html

Reply:
<img width="1127" alt="Screen Shot 2020-11-30 at 5 14 12 PM" src="https://user-images.githubusercontent.com/458432/100584488-8ef78e00-332f-11eb-8398-4525d1ce986d.png">

Reply:
I see :) It's really cool that the compatibility table is built into the type system :)